controllers/signer/interface.go

```diff
@@ -18,10 +18,21 @@ package signer
 
 import (
 	"context"
+	"crypto/x509"
 
 	cmapi "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cert-manager/issuer-lib/api/v1alpha1"
 )
 
-type Sign func(ctx context.Context, cr *cmapi.CertificateRequest, issuerObject client.Object) ([]byte, error)
-type Check func(ctx context.Context, issuerObject client.Object) error
+type Sign func(ctx context.Context, cr CertificateRequestObject, issuerObject v1alpha1.Issuer) ([]byte, error)
+type Check func(ctx context.Context, issuerObject v1alpha1.Issuer) error
+
+type CertificateRequestObject interface {
+	metav1.Object
+
+	GetRequest() (template *x509.Certificate, csr []byte, err error)
+
+	GetConditions() []cmapi.CertificateRequestCondition
+}
```

Comment:
We don't give the Sign function direct access to the `*cmapi.CertificateRequest` or `*certificatesv1.CertificateSigningRequest` object.
Instead, the `template *x509.Certificate` contains all the decoded values from these resources.
`metav1.Object` can be used to get the labels/ annotations of the object.

Reply:
**Question 1:** I find the "interface" solution really elegant! I was trying to find another possible solution e.g. pass the data needed by the Sign and Check funcs every time but the interface solution is just better (and easier to document and test too)

My thinking was to pass the X.509 req straight to the Sign and Check funcs:

```go
type Sign func(ctx, template *x509.Certificate, duration time.Duration, csr []byte, conds []cmapi.CertificateRequestCondition, issuer v1alpha1.Issuer) ([]byte, error)
type Check func(ctx, issuer v1alpha1.Issuer) error
```

Reply:
**Question 2:** with this interface, is it still possible for the user of issuer-lib to update the certificaterequest or certificatesigningrequest with extra conditions or send events with this approach? These two possibilities are important when the external issuer is particularly complex and the implementer wants to give more info at some "intermediate" stages of the Sign and Check func.

Reply:
SetCertificateRequestConditionError can still be used to set custom conditions (https://github.com/cert-manager/issuer-lib/blob/main/controllers/signer/err_set_condition.go)

Reply:
I propose to add a note above the interface to explain that it isn't possible to set conditions "directly", but that it is possible to use `SetCertificateRequestConditionError` as a work around.

What about events?

Reply:
You should be able to set/ send events using the eventRecorder that you keep before you pass it to CombinedController https://github.com/cert-manager/issuer-lib/blob/2427784b3a41157a37a92ca750c7a8242ddac3f0/controllers/combined_controller.go#L49.
You won't be able to create events for the CR resource (because for some reason the event recorder wants a runtime.Object).
However, the issuer-lib will automatically create events. If we want to allow the implementer to create events for CR resources, we can still fix this in a follow-up PR ( without breaking the API ).

Reply:
That makes sense. Let's keep this "feature idea" in mind when we try to use issuer-lib in google-cas-issuer.